02_knight_and_princess/alexey_mesheryakov/graph.cpp

```diff
@@ -0,0 +1,137 @@
+#include <assert.h>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <vector>
+
+constexpr int VERTEX_COUNT = 14;
+constexpr int INVALID_ID = -1;
+
+using std::pair;
+using std::vector;
+using EdgeId = int;
+using VertexId = int;
+
+class Graph {
+ public:
+  class Edge {
+   public:
+    Edge(const pair<VertexId, VertexId>& vertex_ids, const EdgeId& edge_id)
+        : vertex_ids_(vertex_ids), id_(edge_id) {}
+
+    std::string to_string() const {
+      std::stringstream buffer;
+      buffer << "{\"id\":" << id_ << ",\"vertex_ids\":[" << vertex_ids_.first
+             << "," << vertex_ids_.second << "]}";
+      return buffer.str();
+    }
+    pair<VertexId, VertexId> get_vertex_ids() const { return vertex_ids_; }
+
+   private:
+    const EdgeId id_ = INVALID_ID;
+    const pair<VertexId, VertexId> vertex_ids_ = {};
+  };
+  class Vertex {
+   public:
+    explicit Vertex(const VertexId& id) : id_(id) {}
+
+    std::string to_string() const {
+      std::stringstream buffer;
+      buffer << "{\"id\":" << id_ << ",\"edge_ids\":[";
+      for (int i = 0; i < edge_ids_.size() - 1; i++)
+        buffer << edge_ids_[i] << ",";
+      buffer << edge_ids_[edge_ids_.size() - 1] << "]}";
+      return buffer.str();
+    }
+
+    void add_edge(const EdgeId& edge_id) {
+      assert(edge_id_not_exist(edge_id) && "Edge id already exist");
+      edge_ids_.push_back(edge_id);
+    }
+
+   private:
+    bool edge_id_not_exist(const EdgeId& new_edge_id) {
+      bool exist = false;
+      for (const auto& edge_id : edge_ids_)
+        exist = exist || (edge_id == new_edge_id);
+      return !exist;
```

Comment:
Тебе не нужна переменная, можно упростить:
```suggestion
      for (const auto& edge_id : edge_ids_)
        if (edge_id == new_edge_id) {
          return true;
        }
      return false;
```